File: cloexec.c
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>

int main()
{
	if (mkfifo("exit-fifo", 0666) != 0) {
		printf("%s\n", strerror(errno));
		exit(EXIT_FAILURE);
	}
	int fd = open("exit-fifo", O_WRONLY | O_CLOEXEC, 0);
	if (fd < 0) {
		printf("%s\n", strerror(errno));
		exit(EXIT_FAILURE);
	}
	pause();
}

File: cloexec.go
package main

import (
	"log"
	"os"
	"os/signal"
	"syscall"
)

func main() {
	if err := syscall.Mkfifo("exit-fifo", 0666); err != nil && !os.IsExist(err) {
		log.Fatal(err)
	}
	_, err := os.OpenFile("exit-fifo", syscall.O_WRONLY|syscall.O_CLOEXEC, 0)
	if err != nil {
		log.Fatal(err)
	}
	s := make(chan os.Signal)
	signal.Notify(s, syscall.SIGTERM)
	<-s
}
